pkg/cgroup/cgroup_linux.go

```
@@ -0,0 +1,260 @@
+	// controller.
+	cgmProgGetValue = "getvalue"
+)
+
+// Fetch value of the cgroup kernel param from the cgroup manager,
+// if cgroup manager is configured we should just rely on `cgm` cli
+// to fetch all the values for us.
+func getCGroupManagerKernValue(cg CGroup, kernParam string) (limit uint64, err error) {
+	limit = unlimitedMemoryLimit
+
+	cmd := exec.Command(cgmProgName, cgmProgGetValue, cg.Name, cg.Path, kernParam)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	if err = cmd.Run(); err != nil {
+		return limit, err
+	}
+
+	var v uint64
+	v, err = strconv.ParseUint(strings.TrimSpace(out.String()), 10, 64)
+	if err != nil {
+		return limit, err
+	}
+
+	limit = v
+	return limit, nil
+}
+
+// Get cgroup memory limit file path.
+func getCGroupMemoryLimitFilePath(cgPath string) string {
+	path := cgroupMemSysPath
+
+	// Docker generates weird cgroup paths that don't
+	// really exist on the file system.
+	//
+	// For example on regular Linux OS :
+	// `/user.slice/user-1000.slice/session-1.scope`
+	//
+	// But they exist as a bind mount on Docker and
+	// are not accessible :  `/docker/<hash>`
+	//
+	// We we will just ignore if there is `/docker` in the
+	// path ignore and fall back to :
+	// `/sys/fs/cgroup/memory/memory.limit_in_bytes`
+	if !strings.HasPrefix(cgPath, dockerPrefixName) {
+		path = filepath.Join(path, cgPath)
+	}
+
+	// Final path.
+	return filepath.Join(path, memoryLimitInBytes)
+}
+
+// GetCGroupMemoryLimit - Fetches cgroup memory limit either
+// from a file path at '/sys/fs/cgroup/memory', if path fails then
+// fallback to querying cgroup manager.
+func GetCGroupMemoryLimit(pid int) (limit uint64, err error) {
+	limit = unlimitedMemoryLimit
+
+	var pcg ProcCGroup
+	pcg, err = GetProcCGroup(pid)
+	if err != nil {
+		return limit, err
+	}
+
+	memory := pcg.GetCGByName(cgroupMemName)
+	if len(memory) == 0 {
+		return limit, errors.New("Unable to locate any memory limits")
+	}
+
+	// Pick the first entry in memory controller.
+	cg := memory[0]
```

Why ignore others than first entry?  If you are expecting only one entry for memory, you could error out ie multiple memory entires is a unknown situation this may send wrong information.

---

Expectation is only one entry,  error is better.